website/blog/deployers.md

```diff
@@ -0,0 +1,525 @@
+things as concrete as possible, we'll do everything manually the hard way. In
+the next section, we'll leverage Service Weaver's [runtime packages][runtime] to
+make this process a lot easier.
+
+Here's the deployer. See [here][pipes_deployer] for the full source code.
+
+```go
+package main
+
+import ...
+
+func main() {
+    flag.Parse()
+    if err := run(context.Background(), flag.Arg(0)); err != nil {
+        fmt.Fprintln(os.Stderr, err)
+        os.Exit(1)
+    }
+}
+
+func run(ctx context.Context, binary string) error {
+    // Keep reading to see how to implement this function.
+}
+```
+
+Our deployer receives a Service Weaver binary as its first and only argument and
+passes it to a `run` function, which we'll now implement.
+
+```go
+func run(ctx context.Context, binary string) error {
+    // Step 1. Run the binary and establish the pipes between the deployer and
+    // the weavelet.
+    //
+    //  deployer               weavelet
+    // ┌──────────────────┐   ┌───────────────────┐
+    // │ deployerWriter --│---│-> weaveletReader  │
+    // │ deployerReader <-│---│-- weaveletWriter  │
+    // └──────────────────┘   └───────────────────┘
+    weaveletReader, deployerWriter, err := os.Pipe()
+    if err != nil {
+        return err
+    }
+    deployerReader, weaveletWriter, err := os.Pipe()
+    if err != nil {
+        return err
+    }
+
+    cmd := exec.Command(binary)
+    cmd.ExtraFiles = []*os.File{weaveletReader, weaveletWriter}
+    cmd.Env = []string{"ENVELOPE_TO_WEAVELET_FD=3", "WEAVELET_TO_ENVELOPE_FD=4"}
+    if err := cmd.Start(); err != nil {
+        return err
+    }
+
+    // ...
+}
+```
+
+First, we create a pair of pipes. The first pipe runs from `deployerWriter` in
+the deployer to `weaveletReader` in the weavelet. The second runs in the
+opposite direction from `weaveletWriter` in the weavelet to `deployerReader` in
+the deployer. We launch the binary in a subprocess. The `weaveletReader` and
+`weaveletWriter` ends of the pipes will be accessible in the subprocess as file
+descriptors 3 and 4 (see [the exec package][exec] for details). The weavelet
+running in the subprocess opens the file descriptors stored in the
+`ENVELOPE_TO_WEAVELET_FD` and `WEAVELET_TO_ENVELOPE_FD` environment variables to
+establish its connection to the deployer.
+
+```go
+func run(ctx context.Context, binary string) error {
+    // Step 1...
+
+    // Step 2. Send a WeaveletInfo to the weavelet.
+    info := &protos.EnvelopeMsg{
+        WeaveletInfo: &protos.WeaveletInfo{
+            App:           "app",
+            DeploymentId:  uuid.New().String(),
+            Id:            uuid.New().String(),
+            SingleProcess: false,
+            SingleMachine: true,
+        },
+    }
+    if err := protomsg.Write(deployerWriter, info); err != nil {
+        return err
+    }
+
+    // ...
+}
+```
+
+Second, the deployer sends an `EnvelopeMsg` containing a `WeaveletInfo` to the
```

Comment:
I found it slightly confusing to see the explanation after the code (not trivial to figure out whether explanation goes with the code before it or the code after it). Consider ordering as follows:

## 1. Initialize communication with weavelet

The envelope first creates the app process and connects to it via a pair of pipes.

<code>

Explanation as you have.

Reply:
Good point. Done.